server/fishtest/util.py

```diff
@@ -105,62 +109,84 @@ def get_chi2(tasks, bad_users):
     # Now we do the basic chi2 computation.
     df = (rows - 1) * (columns - 1)
     raw_residual = observed - expected
-    chi2 = numpy.sum(raw_residual * raw_residual / expected)
+    ratio = raw_residual ** 2 / expected
+    row_chi2 = numpy.sum(ratio, axis=1)
+    chi2 = numpy.sum(row_chi2)
     p_value = 1 - scipy.stats.chi2.cdf(chi2, df)
 
     # Finally we also compute for each qualifying worker a residual
     # indicating how badly it deviates from the norm.
-    std_error = numpy.sqrt(
-        expected
-        * numpy.outer((1 - row_sums / grand_total), (1 - column_sums / grand_total))
-    )
-    adj_residual = raw_residual / std_error
+    adj_row_chi2 = row_chi2 / (1 - row_sums / grand_total)
+    res_z = scipy.stats.norm.isf(scipy.stats.chi2.sf(adj_row_chi2, columns - 1))
     for idx in range(len(keys)):
-        users[keys[idx]] = numpy.max(numpy.abs(adj_residual[idx]))
+        users[keys[idx]] = max(0, res_z[idx])
 
-    return {"chi2": chi2, "dof": df, "p": p_value, "residual": users}
+    # We compute 95% and 99% thresholds using the Bonferroni correction.
 
+    z_95, z_99 = [scipy.stats.norm.ppf(1 - p / rows) for p in (0.05, 0.01)]
+
+    return {
+        "chi2": chi2,
+        "dof": df,
+        "p": p_value,
+        "residual": users,
+        "z_95": z_95,
+        "z_99": z_99,
+    }
 
-def calculate_residuals(run):
-    bad_users = set()
-    chi2 = get_chi2(run["tasks"], bad_users)
-    residuals = chi2["residual"]
 
-    # Limit bad users to 1 for now
-    for _ in range(1):
+def get_bad_workers(tasks, cached_chi2=None, p=0.001, res=7.0, iters=1):
+    # If we have an up-to-date result of get_chi2() we can pass
+    # it as cached_chi2 to avoid needless recomputation.
+    bad_workers = set()
+    for _ in range(iters):
+        if cached_chi2 is None:
+            chi2 = get_chi2(tasks, exclude_workers=bad_workers)
+        else:
+            chi2 = cached_chi2
+            chached_chi2 = None
```

Comment:
Typo, which might matter for iter>1

Reply:
Thanks. That was indeed a typo. I corrected it.